pymc_marketing/mmm/marginal_effects.py

```diff
@@ -0,0 +1,173 @@
+#   Copyright 2022 - 2025 The PyMC Labs Developers
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+"""Counterfactual sweeps for Marketing Mix Models (MMM)."""
+
+import arviz as az
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+import xarray as xr
+
+
+class CounterfactualSweep:
+    """CounterfactualSweep class is used to perform counterfactual analysis on MMM's."""
+
+    def __init__(
+        self,
+        mmm,
+        X: pd.DataFrame,
+        predictors: list[str],
+        sweep_values: np.ndarray,
+        sweep_type: str = "multiplicative",
+    ):
+        """
+        Initialize and run the counterfactual sweep.
+
+        Parameters
+        ----------
+        - mmm: The marketing mix model instance used for predictions.
+        - X: Original design matrix (DataFrame).
+        - predictors (list[str]): List of predictors to intervene on.
+        - sweep_values (np.ndarray): Array of sweep values.
+        - sweep_type (str): 'multiplicative', 'additive', or 'absolute'.
+            - 'multiplicative': Multiply the original predictor values by each sweep value.
+            - 'additive': Add each sweep value to the original predictor values.
+            - 'absolute': Set the predictor values directly to each sweep value (ignoring original values).
+        """
+        if sweep_type not in ["multiplicative", "additive", "absolute"]:
+            raise ValueError(
+                "sweep_type must be 'multiplicative', 'additive', or 'absolute'."
+            )
+
+        self.mmm = mmm
+        self.X = X
+        self.predictors = predictors
+        self.sweep_values = sweep_values
+        self.sweep_type = sweep_type
+
+        # Run sweep and store results
+        self.run_sweep()
```

Comment:
Not a fan of the running with initialization. In favor of it being two step process

hiding behind the method call like @cetagostini mentioned